### PR TITLE
Remove the need for _UNRELEASED suffix in versions

### LINKED ISSUE #24768: Ensure Version#CURRENT is not `_UNRELEASED` if we are on a release branch
we should make sure that the `Version#CURRENT` is not suffixed with `_UNRELEASED`  if we are on a release branch. I assume we should do this in gradle @rjernst @nik9000 ?

### PR DESCRIPTION
**Pull request by nik9000**
Before this change we used the name of the constant to check if a version was released or not. Versions ending in `_UNRELEASED` were considered unreleased and other versions were considered released. The trouble with this is that we'd often forget to switch a version from unreleased to released. That is a problem because unreleased versions don't get backwards compatibility tests. This changes the logic so that the released versions are inferred from the current version using the following rules:
1. If the current version is a bug fix release all other versions are released.
2. Otherwise, the version constant before the current version and before all the alphas is considered unreleased.
3. If that unreleased version is itself a bug fix release then that is the only unreleased version.
4. Otherwise, the version before it is unreleased.

Some examples:
* Right now in master the current version is 6.0.0-alpha2 which means that the last non-alpha in the list is unreleased. That version is 5.5.0. Because that is a `.0` release the version before that is also considered unreleased. So in master 5.5.0 and 5.4.1 are both considered unreleased.
* Right now in 5.x the current version is 5.5.0 which means that that last non-alpha in the list is unreleased. That version is 5.4.1. It isn't a `.0` release, so the only unreleased version is 5.4.1.
* Right now in 5.4 the current version is 5.4.1. Since that isn't a `.0` release there are no unreleased versions.

This logic is mirrored in both VersionUtils.java and in build.gradle so that gradle can run the appropriate backwards compatibility tests, some of which are already written.

A secondary goal with removing the `_UNRELEASED` suffix is to one day be able to use a script to add the new version constant.

It is also important to note that gradle already has a test that verifies that the list of released versions is correct according to maven central. This means that we won't be able to forget to add a version constant.

Closes #24768

### CODE REVIEW

File: core/build.gradle [hunk 1]
```diff
@@ -305,9 +305,11 @@ task('verifyVersions') {
      * as possible after release. */
     Set<String> actualVersions = new TreeSet<>(
       indexCompatVersions
-        .findAll { false == it.unreleased }
+        .findAll { false == it.snapshot }
         .collect { it.toString() })
 
+    // TODO this is almost certainly going to fail on 5.4 when we release 5.5.0
```
Comment (nik9000) [inline core/build.gradle]: I want to deal with this in a followup. This change is big enough as is.

File: core/src/main/java/org/elasticsearch/Version.java [hunk 2]
```diff
@@ -74,15 +74,17 @@
     public static final Version V_5_3_2 = new Version(V_5_3_2_ID, org.apache.lucene.util.Version.LUCENE_6_4_2);
     public static final int V_5_4_0_ID = 5040099;
     public static final Version V_5_4_0 = new Version(V_5_4_0_ID, org.apache.lucene.util.Version.LUCENE_6_5_0);
-    public static final int V_5_5_0_ID_UNRELEASED = 5050099;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/Version.java]: I almost certainly should have waited to make this change as a purely mechanical followup. It *nasty* to backport. But I already bit this pill so I feel like I have to swallow it now....

File: core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java [hunk 3]
```diff
@@ -308,7 +308,7 @@ public void readFrom(StreamInput in) throws IOException {
         indicesOptions = IndicesOptions.readIndicesOptions(in);
         type = in.readOptionalString();
         source = in.readString();
-        if (in.getVersion().before(Version.V_6_0_0_alpha1_UNRELEASED)) { // TODO change to V_5_3 once backported
+        if (in.getVersion().before(Version.V_6_0_0_alpha1)) { // TODO change to V_5_3 once backported
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java]: This TODO worries me some....

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java]: @jaymode, should this change?

File: core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java [hunk 4]
```diff
@@ -487,7 +487,7 @@ public void readFrom(StreamInput in) throws IOException {
         for (int i = 0; i < size; i++) {
             final String type = in.readString();
             String mappingSource = in.readString();
-            if (in.getVersion().before(Version.V_6_0_0_alpha1_UNRELEASED)) { // TODO change to V_5_3_0 once backported
+            if (in.getVersion().before(Version.V_6_0_0_alpha1)) { // TODO change to V_5_3_0 once backported
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: @jaymode here too?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: And maybe we're missing something in a mixed cluster test? Maybe we just aren't running it yet.....

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java [hunk 5]
```diff
@@ -20,60 +20,106 @@
...
+import java.util.regex.Pattern;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.unmodifiableList;
 
 /** Utilities for selecting versions in tests */
 public class VersionUtils {
+    /**
+     * Matches names of versions that we process. Intentionally skips things like CURRENT.
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: Now that I think about it, maybe I should just skip `CURRENT` and drop the regex entirely.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+public class VersionUtilsTests extends ESTestCase {
+
+    public void testAllVersionsSorted() {
+        List<Version> allVersions = VersionUtils.allReleasedVersions();
+        for (int i = 0, j = 1; j < allVersions.size(); ++i, ++j) {
+            assertTrue(allVersions.get(i).before(allVersions.get(j)));
+        }
+    }
+
+    public void testRandomVersionBetween() {
```
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: I just moved this test and the one above it. They were in the `org.elasticsearch.test.test` package which would have forced me to make `VersionUtils. resolveReleasedVersions` public and that seemed wrong.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        assertEquals(got, Version.V_6_0_0_alpha2);
+
+        // implicit range of one
+        got = VersionUtils.randomVersionBetween(random(), null, VersionUtils.getFirstVersion());
+        assertEquals(got, VersionUtils.getFirstVersion());
+        got = VersionUtils.randomVersionBetween(random(), Version.CURRENT, null);
+        assertEquals(got, Version.CURRENT);
+    }
+
+    static class TestPatchBranch {
```
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: I really wanted to have tests of this logic on the types of versions that we expect to see. This is kind of like the current state of the 5.4 branch.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    public void testResolveReleasedVersionsForPatchBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestPatchBranch.CURRENT, TestPatchBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestPatchBranch.V_5_3_0, TestPatchBranch.V_5_3_1, TestPatchBranch.V_5_3_2, TestPatchBranch.V_5_4_0,
+                TestPatchBranch.V_5_4_1), released);
+        assertEquals(emptyList(), unreleased);
+    }
+
+    static class TestMinorBranch {
```
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: And this one is kind of like 5.x.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    public void testResolveReleasedVersionsForMinorBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestMinorBranch.CURRENT, TestMinorBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestMinorBranch.V_5_3_0, TestMinorBranch.V_5_3_1, TestMinorBranch.V_5_4_0), released);
+        assertEquals(singletonList(TestMinorBranch.V_5_3_2), unreleased);
+    }
+
+    static class TestNextBranch {
```
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: This one is what I imagine 5.x looked like right after we forked the 5.4 branch.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    public void testResolveReleasedVersionsForNextBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestNextBranch.CURRENT, TestNextBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestNextBranch.V_5_3_0, TestNextBranch.V_5_3_1, TestNextBranch.V_5_5_0), released);
+        assertEquals(Arrays.asList(TestNextBranch.V_5_3_2, Version.V_5_4_0), unreleased);
+    }
+
+    static class TestAlphaBranch {
+        public static final Version V_5_3_0 = Version.fromId(Version.V_5_3_0_ID);
```
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: And this one is pretty much like master.

File: core/build.gradle (hunk 1 above)
Comment (rjernst) [inline core/build.gradle]: What do you mean by "fail on 5.4"? The 5.4 branch? I think we need to filter the actualVersions for things less than the current version?

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java [hunk 11]
```diff
@@ -20,60 +20,106 @@
...
-    static {
-        final Field[] declaredFields = Version.class.getFields();
-        final Set<Integer> releasedIdsSet = new HashSet<>();
-        final Set<Integer> unreleasedIdsSet = new HashSet<>();
-        for (final Field field : declaredFields) {
+    /**
+     * Sort versions that have backwards compatibility guarantees from
+     * those that don't. Doesn't actually check whether or not the versions
+     * are released, instead it relies on gradle to have already checked
+     * this which is does in {@code :core:verifyVersions}. So long as the
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: typo: is -> it

File: core/build.gradle [hunk 12]
```diff
@@ -290,7 +290,7 @@ task('verifyVersions') {
     }
     Set<String> knownVersions = new TreeSet<>(xml.versioning.versions.version.collect { it.text() }.findAll { it ==~ /\d\.\d\.\d/ })
 
-    // Limit the known versions to those that should be wire compatible
+    // Limit the known versions to those that should be index compatible
     String currentVersion = versions.elasticsearch.minus('-SNAPSHOT')
     int prevMajor = Integer.parseInt(currentVersion.split('\\.')[0]) - 1
     if (prevMajor == 4) {
```
Comment (rjernst) [inline core/build.gradle]: This logic only needs to exist on the 5.x branches.

File: core/build.gradle [hunk 13]
```diff
@@ -290,7 +290,7 @@ task('verifyVersions') {
     }
     Set<String> knownVersions = new TreeSet<>(xml.versioning.versions.version.collect { it.text() }.findAll { it ==~ /\d\.\d\.\d/ })
```
Comment (rjernst) [inline core/build.gradle]: It is weird to me this task is on core. Could it just be in the root project? And create a precommit task there which depends on this?

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java [hunk 14]
```diff
@@ -20,60 +20,106 @@
...
+        Collections.sort(versions);
+        assert versions.get(versions.size() - 1).equals(current) : "The highest version must be the current one "
+            + "but was [" + versions.get(versions.size() - 1) + "] and current was [" + current + "]";
+
+        if (current.revision != 0) {
+            /* If we are in a patch release there should be no unreleased version constants.
+             * If there are, gradle will yell about it. */
+            return new Tuple<>(unmodifiableList(versions), emptyList());
+        }
+        /* If we are on a patch release then we know that at least the version before the
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: patch release -> stable branch?

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java [hunk 15]
```diff
@@ -20,60 +20,106 @@
...
+         * current one is unreleased. If it is released then gradle would be complaining. */
+
+        /* Technically we don't support backwards compatiblity for alphas, betas,
+         * and rcs. But the testing infrastructure requires that we act as though we
+         * do. This is a difference between the gradle and Java logic but should be
+         * fairly safe as it is errs on us being more compatible rather than less....
+         * Anyway, the upshot is that we never declare alphas as unreleased, no
+         * matter where they are in the list */
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: except for CURRENT on master...

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+        // implicit range of one
+        got = VersionUtils.randomVersionBetween(random(), null, VersionUtils.getFirstVersion());
+        assertEquals(got, VersionUtils.getFirstVersion());
+        got = VersionUtils.randomVersionBetween(random(), Version.CURRENT, null);
+        assertEquals(got, Version.CURRENT);
+    }
+
+    static class TestPatchBranch {
+        public static final Version V_5_3_0 = Version.fromId(Version.V_5_3_0_ID);
```
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Using constants from `Version` makes removing old constants time consuming. Can you use `Version.fromString` throughout these tests?

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 9 above)
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Can you rename this `TestStableBranch`?

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 7 above)
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Can you rename this `TestReleaseBranch`?

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    public void testResolveReleasedVersionsForNextBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestNextBranch.CURRENT, TestNextBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestNextBranch.V_5_3_0, TestNextBranch.V_5_3_1, TestNextBranch.V_5_5_0), released);
+        assertEquals(Arrays.asList(TestNextBranch.V_5_3_2, Version.V_5_4_0), unreleased);
+    }
+
+    static class TestAlphaBranch {
```
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Can you rename this `TestUnstableBranch`?

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        assertEquals(Arrays.asList(TestAlphaBranch.V_5_3_2, TestAlphaBranch.V_5_4_0), unreleased);
+    }
+
+    /**
+     * Test that the released inference logic produces expected "unreleased"
+     * versions. This test will fail every time we create a new version
+     * constant and it should be upgraded by hand. Take this opportunity to
+     * make sure that it lines up with reality.
+     */
+    public void testUnreleaseCurrent() {
```
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: I don't think we should have a test that fails every time we create a new version; that makes automating version bumping more difficult. Can we instead test that CURRENT is equal to the last unreleased version?

**Review by rjernst — COMMENTED** [review]
I left some comments.

File: core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java (hunk 3 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java]: Yes it should change. I'll open a PR.

File: core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java (hunk 4 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: This should also change. We do not catch this in our BWC/mixed cluster tests since these checks just guard the need to do content type auto-detection. In versions > 5.3.0 we always send JSON so we do not need to autodetect, but this check being wrong kept the auto detection in place when it wasn't necessary.

File: core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java (hunk 3 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/action/admin/indices/mapping/put/PutMappingRequest.java]: Opened #24835 for both

File: core/build.gradle (hunk 1 above)
Comment (nik9000) [inline core/build.gradle]: I believe so. I added this here because I noticed it here. I think it'd be nice to fix it in a separate PR though because it isn't really related to this.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 9 above)
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Sure!

File: core/build.gradle (hunk 12 above)
Comment (nik9000) [inline core/build.gradle]: Yes. I figure it helps to keep them in sync so backporting is easier.

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java (hunk 15 above)
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: I think this is leftovers. Something like this PR will help us level set how we deal with these and then we can clean them up from there. Like, we can remove this funny exception for alphas in a followup, tweaking all the tests that need it.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 16 above)
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Yeah, that is better.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 18 above)
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: I'll drop it. I had thought it'd be a nice reminder for the next few releases *until* we automate the version bumping.

File: test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java (hunk 14 above)
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/VersionUtils.java]: Sure. That is more descriptive.

File: core/build.gradle (hunk 13 above)
Comment (nik9000) [inline core/build.gradle]: Rereading, yes, that looks like what you'd prefer. I'll do that now.

**Comment by nik9000** [comment]
@rjernst I think this should be ready for another round.

**Comment by rjernst** [comment]
The changes look fine, but I still see the old test class names? I think we should keep with consistent naming, meaning using `unstable` for `master`, `stable` for `N.x`, and `release` (no `d` on the end like I see in one of the class names) for `N.M`.  Albeit I am copying this naming from Lucene, but I have personally used this terminology to talk about ES branching to other engineers at elastic since I've been here, so I think we standardize on it if there is any actual question about it?

**Comment by nik9000** [comment]
Ok. I understand the naming convention now. I like it. I'll switch.

**Comment by nik9000** [comment]
@rjernst renamed.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+    public void testResolveReleasedVersionsForUnstableBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestUnstableBranch.CURRENT,
+                TestUnstableBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestUnstableBranch.V_5_3_0, TestUnstableBranch.V_5_3_1, TestUnstableBranch.V_5_5_0), released);
+        assertEquals(Arrays.asList(TestUnstableBranch.V_5_3_2, Version.V_5_4_0), unreleased);
+    }
+
+    static class TestUnstableBranchWithAlphas {
```
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: I don't think this needs "WithAlphas", because the unstable branch always has alphas (when we switch to beta, we cut a new stable branch).

**Review by rjernst — APPROVED** [review]
LGTM, thanks for the naming! I left one last minor comment.

File: test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java (hunk 19 above)
Comment (nik9000) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: Do you think I should drop the test without alphas then? Or rename it? I'd kind of like to keep it because it is telling is only one of the two tests fails.

Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java]: We should only have one unstable test, and that should be how the unstable branch Version class always looks, ie with CURRENT being an alpha. I'm not sure I understand the testing concern, you have tests here for each of the patterns we handle on each branch, so we should be covered?
